ee/tabby-webserver/src/oauth/general.rs

```diff
@@ -0,0 +1,205 @@
+use std::sync::{Arc, Mutex};
+
+use anyhow::{anyhow, Result};
+use async_trait::async_trait;
+use serde::Deserialize;
+use tabby_schema::auth::{AuthenticationService, OAuthCredential, OAuthProvider};
+
+use cached::proc_macro::cached;
+use cached::TimedCache;
+
+use super::OAuthClient;
+use crate::bail;
+
+#[derive(Debug, Deserialize)]
+#[allow(dead_code)]
+struct GeneralOAuthResponse {
+    #[serde(default)]
+    access_token: String,
+    #[serde(default)]
+    scope: String,
+    #[serde(default)]
+    token_type: String,
+
+    #[serde(default)]
+    error: String,
+    #[serde(default)]
+    error_description: String,
+    #[serde(default)]
+    error_uri: String,
+}
+
+#[derive(Debug, Deserialize, Clone)]
+#[allow(dead_code)]
+struct GeneralUserInfo {
+    email: String,
+    name: String,
+}
+
+pub struct GeneralClient {
+    client: reqwest::Client,
+    auth: Arc<dyn AuthenticationService>,
+    user_info: Mutex<Option<GeneralUserInfo>>
+}
+
+#[derive(Debug, Deserialize, Clone)]
+pub struct OAuthConfig {
+    authorization_endpoint: String,
+    token_endpoint: String,
+    userinfo_endpoint: String,
+    scopes_supported: Vec<String>,
+}
+
+impl GeneralClient {
```

Comment:
I would like to add some tests to this class but it heavily relies on the DB (via `read_credential`) and outgoing requests via `reqwest::Client`. 

I am unsure on how to mock or intercept these calls and would appreciate the assistance here.